Cognite.Extensions/TimeSeries/TimeSeriesExtensions.cs

```diff
@@ -436,5 +436,91 @@ Task<IEnumerable<TimeSeriesCreate>> asyncBuildTimeSeries(IEnumerable<string> ids
 
             return new CogniteResult<TimeSeries, TimeSeriesUpdateItem>(errors, null);
         }
+
+        /// <summary>
+        /// Insert or update a list of timeseries, handling errors that come up during both insert and update.
+        /// Only timeseries that differ from timeseries in CDF are updated.
+        /// 
+        /// All given timeseries must have an external id, so it is not in practice possible to use this to change
+        /// the externalId of timeseries.
+        /// 
+        /// Timeseries are returned in the same order as given.
+        /// </summary>
+        /// <param name="tss">TimeSeries resource</param>
+        /// <param name="upserts">Assets to upsert</param>
+        /// <param name="chunkSize">Number of asset creates, retrieves or updates per request</param>
+        /// <param name="throttleSize">Maximum number of parallel requests</param>
+        /// <param name="retryMode">How to handle retries on errors</param>
+        /// <param name="sanitationMode">How to sanitize creates and updates</param>
+        /// <param name="options">How to update existing assets</param>
+        /// <param name="token">Cancellation token</param>
+        /// <returns>Result with failed creates/updates and list of assets</returns>
+        /// <exception cref="ArgumentException">All upserted assets must have external id</exception>
+        public static async Task<CogniteResult<TimeSeries, TimeSeriesCreate>> UpsertAsync(
```

Comment:
both upserts are almost identical except for a couple field and method names... if this is being extended to all resource types, consider implementing an interface (eg. with `CreateOrUpdate()` and any other needed method or property) and extending it instead of each resource.

Reply:
Problem is that the interface needs to be defined on something, and here all methods are extension methods to the resources from the SDK, so doing this means passing the methods as arguments, which we could do, but it still ends up being a bit of a pain. I'll see if I can do it. I agree that it results in a lot of duplicated code.

Reply:
Yeah. Signature of the method would look like this:

```
public static async Task<CogniteResult<TResult, TCreate>> UpsertAsync<TResult, TCreate, TUpdate>(
            IEnumerable<TCreate> upserts,
            int chunkSize,
            int throttleSize,
            RetryMode retryMode,
            SanitationMode sanitationMode,
            UpsertParams? options,
            Func<TCreate, string> getExternalId,
            Func<IEnumerable<string>,
                Func<IEnumerable<string>,IEnumerable<TCreate>>,
                int,
                int,
                RetryMode,
                SanitationMode,
                CancellationToken,
                Task<CogniteResult<TResult, TCreate>>> getOrCreate,
            Func<IEnumerable<TUpdate>, int, int, RetryMode, SanitationMode, CancellationToken, CogniteResult<TResult, TUpdate>> update,
            Func<TCreate, TUpdate> getUpdate,
            CancellationToken token
            )
```

Which IMO is too much. This doesn't help with maintainability or readability.

Reply:
Really this question goes for all the extension methods. The issue is that the API itself is just too inconsistent, and the SDK aims to just expose the SDK and do nothing else.